walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,185 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.List;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    //If there are no events or no one is attending the event, 
+    // all requests are possible if they are within the range of the whole day.
+
+    // The basic functionality of optional attendees is that if one or more time slots exists 
+    // so that both mandatory and optional attendees can attend, return those time slots. 
+    // Otherwise, return the time slots that fit just the mandatory attendees.
+
+    // Find all events that attendees in the meeting request are already booked for.
+
+    // Mandatory attendees.
+    List<Event> mandatoryBookedEvents = findBookedEvents(events, request.getAttendees());
+   
+    // Optional attendes.
+    List<Event> optionalBookedEvents = findBookedEvents(events, request.getOptionalAttendees());
+
+    // All attendees - Mandatory + Optional.
+    List<Event> allBookedEvents = getAllUniqueEvents(mandatoryBookedEvents, optionalBookedEvents);
+    allBookedEvents = sortEventsByStartTime(allBookedEvents);
+
+    // No possible meeting if request is longer than a day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+    }
+    // No events/no attendess means meeting can be anytime.
+    if(events.isEmpty() || (mandatoryBookedEvents.isEmpty() && optionalBookedEvents.isEmpty())) {
```

Comment:
Is it possible for the mandatory and optional events to be non-empty if the events list is empty? or visa versa?
(To ask this question another way are you just doing the same check twice?)

Reply:
Done